### PR TITLE
Golang type refactor

### PR DESCRIPTION
**Pull request by prestwich**
We defined types for RawHeader and Hash256Digest. This PR adds them to all relevant functions

### CODE REVIEW

**Comment by prestwich** [comment]
@ErinHales tests needed for these functions in `types.go`

```
NewHash160Digest
NewHash256Digest
NewRawHeader
HeaderFromRaw
HeaderFromHex
```

**Comment by ErinHales** [comment]
Wait, I don't see a `NewHash160Digest` function.

**Comment by prestwich** [comment]
uhhh I forget to push that commit

File: golang/btcspv/bitcoin_spv.go [hunk 1]
```diff
@@ -338,18 +349,20 @@ func ValidateVout(vout []byte) bool {
...
+	return res
 }
 
 // ExtractMerkleRootBE returns the transaction merkle root from a given block header
 // The returned merkle root is big-endian
-func ExtractMerkleRootBE(header []byte) []byte {
-	return ReverseEndianness(ExtractMerkleRootLE(header))
+func ExtractMerkleRootBE(header RawHeader) Hash256Digest {
+	LE := ExtractMerkleRootLE(header)
+	return ReverseHash256Endianness(LE)
```
Comment (ErinHales) [inline golang/btcspv/bitcoin_spv.go]: Not a big deal, but we could think about making this into a 1-liner again `return ReverseHash256Endianness(ExtractMerkleRootLE(header))`

File: golang/btcspv/bitcoin_spv.go [hunk 2]
```diff
@@ -377,35 +390,37 @@ func CalculateDifficulty(target sdk.Uint) sdk.Uint {
...
+	return res
 }
 
 // ExtractPrevBlockHashBE returns the previous block's hash from a block header
 // Returns the hash as a big endian []byte
-func ExtractPrevBlockHashBE(header []byte) []byte {
-	return ReverseEndianness(ExtractPrevBlockHashLE(header))
+func ExtractPrevBlockHashBE(header RawHeader) Hash256Digest {
+	LE := ExtractPrevBlockHashLE(header)
+	return ReverseHash256Endianness(LE)
```
Comment (ErinHales) [inline golang/btcspv/bitcoin_spv.go]: Same here: `ReverseHash256Endianness(ExtractPrevBlockHashLE(header))`

File: golang/btcspv/bitcoin_spv_test.go [hunk 3]
```diff
@@ -100,8 +121,14 @@ func preprocessObject(m map[string]interface{}) {
...
-			// overwrite the string with a []byte
-			m[k] = DecodeIfHex(v.(string))
+			buf := DecodeIfHex(v.(string))
+			if len(buf) == 32 {
+				m[k], _ = NewHash256Digest(buf)
+			} else if len(buf) == 80 {
+				m[k], _ = NewRawHeader(buf)
+			} else {
+				m[k] = buf
+			}
```
Comment (ErinHales) [inline golang/btcspv/bitcoin_spv_test.go]: I noticed this block of code is repeated a lot, could it be extracted into another function?

File: golang/btcspv/validate_spv_test.go [hunk 4]
```diff
@@ -96,23 +111,23 @@ func (suite *UtilsSuite) TestParseHeader() {
...
+	// 	digest, version, prevHash, merkleRoot, timestamp, target, nonce, err := ParseHeader(input)
+	// 	suite.Nil(digest)
+	// 	suite.Equal(version, uint(0))
+	// 	suite.Nil(prevHash)
+	// 	suite.Nil(merkleRoot)
+	// 	suite.Equal(timestamp, uint(0))
+	// 	suite.Equal(target, sdk.NewUint(0))
+	// 	suite.Equal(nonce, uint(0))
+	// 	suite.EqualError(err, expected)
+	// }
```
Comment (ErinHales) [inline golang/btcspv/validate_spv_test.go]: Delete this

**Pull request merged**